docs/programmers_guide/witness_formal_spec.md

```diff
@@ -0,0 +1,222 @@
+# Block Witness Formal Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## The Stack
+
+Witnesses are executed on a stack machine that builds tries/calculates hashes.
+
+The stack consists of pairs `(node, hash)`.
```

Comment:
Nodes are RLP strings and hashes are 32 bytes, correct?

Reply:
no, nodes are objects in that case already, and hashes are 32 bytes, yes